### PR TITLE
DAO-1524 [Mobile] Proposal List Page - Filters overlay

### PR DESCRIPTION
**Pull request by TravellerOnTheRun**
# Filter System Refactoring: From Validate Functions to Centralized Filtering

## 📋 Summary

Implementing a **hybrid filtering approach** that replaces the old `validate` function pattern with a more efficient, type-safe filtering system.

### **❌ Old Approach (Main Branch): Validate Functions**
```typescript
// Each filter had its own validate function
export interface FilterItem {
  id: string
  type: FilterType
  label: string
  value: string
  validate(proposal: Proposal): boolean  // ← This was the problem
}

// Filtering logic was scattered across validate functions
const filterProposals = (proposals: Proposal[], activeFilters: FilterItem[]) => {
  return proposals.filter(proposal => {
    return filterGroups.every(filters => 
      filters.some(f => f.validate(proposal))  // ← Called validate on each proposal
    )
  })
}
```

### **✅ New Approach: Centralized Filtering Logic**
```typescript
// Clean interface without validate functions
export interface FilterItem {
  id: string
  type: FilterType
  label: string
  value: string
  isAll?: boolean
  exclusive?: boolean
  // No more validate function!
}

// Centralized filtering logic
const filterProposals = (proposals: Proposal[], activeFilters: FilterItem[]) => {
  return proposals.filter(proposal => {
    // Direct filtering logic - no function calls
    const searchFilters = activeFilters.filter(f => f.type === FilterType.SEARCH)
    const categoryFilters = activeFilters.filter(f => f.type === FilterType.CATEGORY)
    // ... etc
  })
}
```

#### **Old Approach (Validate Functions)**
```typescript
// Each proposal had to go through multiple function calls
proposal => {
  return filterGroups.every(filters => 
    filters.some(f => f.validate(proposal))  // Function call overhead
  )
}

// Each validate function created a new execution context
const createCategoryFilter = (value: string): FilterItem => ({
  validate: (proposal: Proposal) => {
    return proposal.category?.toLowerCase().includes(value.toLowerCase())
  }
})
```

#### **New Approach (Direct Filtering)**
```typescript
// Direct property access with optimized logic
proposal => {
  // Group filters by type once, then apply logic directly
  const categoryFilters = activeFilters.filter(f => f.type === FilterType.CATEGORY && f.value)
  if (categoryFilters.length > 0) {
    const categoryMatch = categoryFilters.some(categoryFilter => {
      const filterValue = categoryFilter.value
      return proposal.category?.toLowerCase()?.includes(filterValue.toLowerCase())
    })
    if (!categoryMatch) return false
  }
  // ... continue with other filter types
}
```

**Why This Is Better:**
- **No Function Call Overhead**: Direct property access instead of function calls
- **Grouped Operations**: Process all filters of the same type together instead of individually
- **Early Exit Strategy**: Stops processing as soon as a filter fails
- **Reduced Memory Allocation**: No new execution contexts for each validate function

#### **Old Approach Problems:**
```typescript
// Scattered validation logic - hard to maintain
const createCategoryFilter = (value: string): FilterItem => ({
  validate: (proposal: Proposal) => {
    return proposal.category?.toLowerCase().includes(value.toLowerCase())
  }
})

const createStatusFilter = (value: ProposalState): FilterItem => ({
  validate: (proposal: Proposal) => {
    return proposal.proposalState === value
  }
})

// Each filter type had its own validation logic
// Hard to debug when something goes wrong
// No centralized error handling
```

#### **New Approach Benefits:**
```typescript
// Centralized, type-safe filtering logic
const filterProposals = (proposals: Proposal[], activeFilters: FilterItem[]) => {
  return proposals.filter(proposal => {
    // All filtering logic in one place
    // Easy to debug and modify
    // Consistent error handling
    // Type-safe property access
  })
}
```

**Why This Is Better:**
- **Single Source of Truth**: All filtering logic in one function
- **Easier Debugging**: Can set breakpoints and inspect the entire filtering process
- **Consistent Error Handling**: Single place to handle edge cases
- **Type Safety**: TypeScript can verify property access at compile time
- **Easier Testing**: Can test the entire filtering logic as one unit

### **3. Responsive Design Integration**
- **Before**: Desktop-only sidebar approach
- **After**: Hybrid modal/sidebar that adapts to screen size
- **Result**: Better mobile experience with full-screen modal

### **4. Enhanced User Experience**
- **Before**: Basic filter chips with limited functionality
- **After**: Smart filter management with "All" filter restoration
- **Result**: More intuitive filtering behavior

### **Filter State Management**
```typescript
// New centralized hook
export const useProposalFilters = (): FilterState & FilterActions => {
  // Handles all filter logic in one place
  // Smart "All" filter restoration
  // Exclusive filter handling (time filters)
  // Search integration
}
```

### **Responsive Filter Sidebar**
```typescript
// Hybrid approach: Modal on mobile, Sidebar on desktop
export function FilterSideBar({ isOpen, onClose, ... }) {
  const isDesktop = useIsDesktop()
  
  if (isDesktop) {
    return <SidebarContent />  // Desktop: persistent sidebar
  }
  
  return <Modal><SidebarContent /></Modal>  // Mobile: full-screen modal
}
```

### **Smart Filter Logic**
- **Exclusive Filters**: Time filters work as radio buttons (only one at a time)
- **"All" Filter Restoration**: Automatically restores "All" when no specific filters are active
- **Milestone Handling**: Special logic for milestone-related category filters
- **Search Integration**: Seamless combination of search and other filters

## 📊 Performance Comparison

| Aspect | Old (Validate Functions) | New (Hybrid Approach) |
|--------|-------------------------|----------------------|
| **Function Calls** | Multiple function calls per proposal | Direct property access |
| **Memory Usage** | Higher - function call overhead | Lower - direct operations |
| **Type Safety** | Limited - validate functions could return anything | Full - TypeScript enforces correct types |
| **Maintainability** | Scattered logic across multiple functions | Centralized in one place |
| **Mobile UX** | Desktop-only sidebar | Responsive modal/sidebar |

### **Mobile Experience**
- **Before**: Desktop sidebar that didn't work well on mobile
- **After**: Full-screen modal with proper mobile UX
- **Result**: Better touch interaction and visual hierarchy

### **Filter Management**
- **Before**: Basic filter chips with limited functionality
- **After**: Smart filter chips with proper categorization
- **Result**: Clear visual separation between search and other filters

### **Responsive Design**
- **Before**: One-size-fits-all approach
- **After**: Adaptive design that responds to screen size
- **Result**: Optimal experience on all devices

### **Enhanced Architecture**
- ✅ Centralized filtering logic in `filterProposals` function
- ✅ Type-safe filter management with proper interfaces
- ✅ Responsive design with `useIsDesktop` hook
- ✅ Smart filter state management

## 🚀 Future Benefits

This hybrid approach provides:
- **Easy Extension**: Adding new filter types is straightforward
- **Better Testing**: Centralized logic is easier to test
- **Performance**: Optimized for large datasets
- **Consistency**: Same pattern can be applied to other tables

---

**Key Achievement**: Successfully resolved the rebase conflict by implementing a **hybrid filtering approach** that combines the best of both worlds - the type safety of the new system with the responsive design needed for mobile users, while eliminating the performance overhead of the old `validate` function approach.

### CODE REVIEW

File: src/app/proposals/components/filter/types.ts [hunk 1]
```diff
@@ -14,7 +12,6 @@ export interface FilterItem {
   value: string
   isAll?: boolean // ex: All categories, All statuses, All proposals, etc.
   exclusive?: boolean // if true, it cannot be selected with other filters of the same type
-  validate(proposal: Proposal): boolean
```
[the remark this answers is not in the document]
Comment (TravellerOnTheRun) [reply src/app/proposals/components/filter/types.ts:17]: I will provide a detailed description. It's supposed to be faster, than previous approach. Just to note, mine was created earlier, I evaluated 1523 and compared the 2, this is the result.

File: src/app/proposals/components/filter/useProposalFilters.tsx [hunk 2]
```diff
@@ -1,24 +1,45 @@
...
+      }
+      // If new filter is exclusive (like time filters), remove other filters of same type
+      if (newFilter.exclusive) {
         return prev.filter(f => f.type !== newFilter.type).concat(newFilter)
       }
-      // else, add the new filter and remove any "all" filter of the same type
+
+      // Special handling for milestone filters
+      if (newFilter.type === FilterType.CATEGORY) {
+        if (newFilter.value === MILESTONE_SEPARATOR) {
```
[the remark this answers is not in the document]
Comment (TravellerOnTheRun) [reply src/app/proposals/components/filter/useProposalFilters.tsx:23]: It's not logical to have both "Grant - all miltestones" and "Grant - Milestone 1(2,3)", thi sis what this piece of code is for, do you have a better way of solving this? Note: choosing 'Grants - Milestone 1' and 'Grant - Milestione 2(or3)' works.

File: src/app/proposals/components/filter/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (TravellerOnTheRun) [reply src/app/proposals/components/filter/types.ts:17]: If you think(know), that performance wise, your will be better, let's take it into Slack and discuss

File: src/app/proposals/components/filter/FilterSideBar.tsx [hunk 3]
```diff
@@ -1,63 +1,187 @@
...
+          />
+        </li>
+        {options.map((option, i) => (
+          <li key={i}>
+            <SelectableItem
+              selected={isFilterSelected(option, type)}
+              option={option}
+              onClick={() => handleFilterToggle(option, type)}
+              data-testid={`${type}Filter-${option.label}`}
+              variant="round"
```
[the remark this answers is not in the document]
Comment (TravellerOnTheRun) [reply src/app/proposals/components/filter/FilterSideBar.tsx:110]: I just saw the design and for some reason it's all `round`, is it verified by design? If so, I will do that, no problem.

File: src/app/proposals/components/filter/FilterSideBar.tsx [hunk 4]
```diff
@@ -1,63 +1,187 @@
...
-      onAddFilter(option)
+      // Create appropriate filter based on type
+      const filter: FilterItem = {
+        id: `${type}-${option.value}-${Date.now()}`,
+        type,
+        label: option.label,
+        value: option.value,
+        exclusive: type === FilterType.TIME, // Time filters are exclusive
+      }
+      onAddFilter(filter)
```
[the remark this answers is not in the document]
Comment (TravellerOnTheRun) [reply src/app/proposals/components/filter/FilterSideBar.tsx:51]: Yeah, but it's a filter toggle for all the types of filters. So one type won't work, but maybe I can create the same function to create appropriate filter based on type.

**Comment by rodrigoncalves** [comment]
> #### **Old Approach (Validate Functions)**
> ```ts
> // Each proposal had to go through multiple function calls
> proposal => {
>   return filterGroups.every(filters => 
>     filters.some(f => f.validate(proposal))  // Function call overhead
>   )
> }
> 
> // Each validate function created a new execution context
> const createCategoryFilter = (value: string): FilterItem => ({
>   validate: (proposal: Proposal) => {
>     return proposal.category?.toLowerCase().includes(value.toLowerCase())
>   }
> })
> ```
- This old approach encodes the rule declaratively: "at least on filter per group". `every`+`some` communicates intent better than a long imperative block.
- Logic stays co-located with each filter. Lower coupling.
- Short-circuit already happens in both `every` and `some`. Early exists are equivalent to explicit `if (!match) return false`.
- Function call overhead isn’t really an issue here. Modern JS engines inline small functions like `validate`, so the cost is close to zero. 
- This approach retains per-type modularity and testability.
- Maintainability favors modular validators and typed dispatch, not a monolith.

> ## 🚀 Future Benefits
> This hybrid approach provides:
> 
> * **Easy Extension**: Adding new filter types is straightforward

Adding a new filter is easier when you add a new validator and a small test, not when you edit a large conditional block.

> * **Better Testing**: Centralized logic is easier to test

Small, focused validators make tests simpler and faster to run.

> * **Performance**: Optimized for large datasets
> * **Consistency**: Same pattern can be applied to other tables

The `every`+`some` law is a clean, reusable pattern: “one-of per group, all groups required.”
